### DISCUSSION TITLE
Guide: How to install Termux on the External Storage (Android Adoptable Storage)

### DISCUSSION THREAD

**Discussion by HaruYou27** [termux/termux-app#3567] [General]
# Requirement
1. Root access
3. Android shell access (adb shell or terminal emulation app)
4. SD card.
# Step by step
1. You need to format your SD card as _Internal Storage_. Make sure to backup your data on the SD card before doing that.
2. Follow this [guide](https://support.google.com/android/answer/12153449?hl=en) on google support to format your SD card.
3. Go to _Settings -> Developer Options_ and enable _Force allow apps on external_ to forcibly move Termux to SD card.
4. Go to Settings -> Apps -> Termux -> Storage -> Move to SD card or Change storage or whatever similar option.
5. Now use a **Root** file explorer and go to `mnt/expand/`
6. You should see a folder here with a name like `60a327d3-d9af-49ca-99af-ff98f7ec86b0`, that's where your SD card is mounted. (The name will be different but it's kinda look like that)
7. Go into that folder and locate `com.termux` folder in `user/` 
8. Finally, you need to create symlink in `/data/data/` because Termux is hardcoded to that location.
> I use adb shell with root access, but an Android terminal emulator with root access might also works.
9. Plug your phone to your PC, I assume that you know how to setup adb.
10. `adb root` to get root access
11. `adb shell` to open Android shell
12. `cd /data/data` to go to where Termux expected.
13. `ln -s <path to your com.termux folder>` In my case it's `ln -s /mnt/expand/60a327d3-d9af-49ca-99af-ff98f7ec86b0/user/0/com.termux`
14. Open your Termux app and see if it's working fine.

**Comment by Biswa96** [comment]
If anyone is using this method with LineageOS termux may not connect with network after reboot. In that case, go to Settings > Apps > Termux > mobile data & Wi-Fi > toggle 'Unrestricted data usage' on and off. Now termux can connect with network. Follow those above steps after every reboot.

The issue and workaround was reported here https://gitlab.com/LineageOS/issues/android/-/issues/6444

**Comment by Musheer360** [comment]
I was able to install it on the SD card, but after each reboot Termux get reset, and it installs bootstrap packages again.

**Reply by HaruYou27** [reply]
> I was able to install it on the SD card, but after each reboot Termux get reset, and it installs bootstrap packages again.

Did Termux re-install bootstrap packeges on your internal storage ? Some Android OS check all apps on boot and _fix_ any abnormally.

**Comment by shiumano** [comment]
Alternatives that do not require root access
https://github.com/termux/termux-app/discussions/3567#discussioncomment-11416142

**Reply by Biswa96** [reply]
"but in my environment, the linker gave an error and stopped working." - do you mean it does not work?

**Reply by shiumano** [reply]
Yes, I don't remember much better now that I don't use this method anymore, but I do remember that there was an error in some binary.
I moved all the `/data/data/com.termux/files` and created a symbolic link, but it stopped working properly.

It may work fine. Even if it didn't work, it would almost never be a problem with proot.

**Comment by kuskebabi** [comment]
also root file manager isnt needed. as you can access filesystem from adb shell itself. you just need to cd/ls to the path at sdcard and copy it, then do the command at /data/data

**Reply by astroskyoffical** [reply]
Can you explain this more? I don't have root

**Reply by Biswa96** [reply]
~~You need root permission to begin this journey.~~

**Reply by agnostic-apollo** [reply]
You don't need root and its not recommended to create directories under `/data/data` for apps that don't exist. Have told you this [before](https://matrix.to/#/!rrHhKvLYtBVDOISphu:gitter.im/$JayjUCc00_MXEE8kOxCMDURDc5f2Ta3KvwPoeUwernc?via=gitter.im&via=matrix.termux.club&via=matrix.org) on matrix.

All you need to do is install a Termux plugin that has same `sharedUserId` as `termux-app`, like Termux:Boot and install that plugin app on external storage and Termux app on internal storage. Then run `adb shell "pm dump com.termux.boot | grep 'dataDir='"
` to find `dataDir=` path of the plugin app from pc using [`adb`](https://developer.android.com/tools/adb). Then open a [failsafe session](https://wiki.termux.com/wiki/Recover_a_broken_environment) in Termux and run `rm -rf "/data/data/com.termux/files"` and then `ln -s "<data_dir_path_of_plugin/termux/rootfs>" "/data/data/com.termux/files"`, where `<data_dir_path_of_plugin/termux/rootfs>` is something like `/mnt/expand/VVVVVVVV-VVVV-VVVV-VVVV-VVVVVVVVVVVV/user/0/com.termux.boot/termux/rootfs`, and then restart Termux app so that bootstrap gets installed.

You can also find the data directory path of the plugin from in-app Termux settings about page under `FILES_DIR` heading in `v0.119.0-beta.1`, and remove the `/files` suffix, so `adb`/root isn't needed either. Maybe a plugin that automatically does this can be created too.

**Reply by atikattar1104** [reply]
Is It Possible To Have The Bootstraped And Functional RootFS Moved To A User Accessible Location (E.g. `/storage/emulated/0/Termux_RootFS/`, Or Etc.), And Then Create A Symlink At Termux's Hardcoded Location, Pointing To That Moved RootFS Path? 

If Possible, The Termux App Itself Would Be Equivalent To A Computer Which Supports Live Booting, And The Moved RootFS Would Be Equivalent To A Portable Live ISO With Persistency. 

This Would Also Allow Adding Encryption To The RootFS, Considering A Large Number Of Enthusiasts Have Worked On Creating Vault Apps With Various Encryption Methods And Algorithms Support Like Veracrypt (Apps :- AndroidCrypt, Arcanum, EDS Lite, Etc.), LUKS (Apps :- EDS Lite, Etc.), GoCryptFS And CryFS (Apps :- DroidFS, Etc.), And Various Other Encryption Methods And Algorithms. 

If Possible, This Would Make Termux, A Linux Subsystem For Android, That Is Highly Secure And Protected Against Surveillance.

**Reply by robertkirkman** [reply]
> `/storage/emulated/0/Termux_RootFS/`

unfortunately, this location does not work unless the partition filesystem type of `/storage/emulated` is changed from sdcardfs to something else, which requires root and probably recompiling the ROM with patches.

**Reply by atikattar1104** [reply]
Yeah. `sdcardfs` Is A Restrictive Filesystem, Only Allowing CRUD Ops From File Management Perspective. This Can't Work. 

There Are Some Veracrypt Based Vault Creation Softwares That Require The Created Vault To Be Formatted To FAT Or exFAT. I Think Veracrypt Itself Needs These Formats To Work. Would Any Of These Work?

Also, I Found Out That The Filesystem Formats Used In `/data` Path, Which Are `EXT4` For Low End Phones, And `F2FS` For Modern High End Flagship Phones. Since None Of The Vaults Support A FS Format Like These, Veracrypt Based Approach Would Likely Fail. 

BUT, If We Use Apps Like Anemo And Aer, Which Use The Same `/data` Path To Store Data, Just Without Powerful Encryption Algorithms, But Rather Password And Biometric-Based Storage Location Access, And Use DocumentsUI/SAF's Path For Anemo's And Aer's Mounted Path As Symlink Destination For The Symlink Created At Termux Hardcoded Location `/data/data/com.termux/`, We Should Be Able To (THEORETICALLY) Boot Termux Environment From This New Path.

**Reply by atikattar1104** [reply]
NVM, You Cannot Create A Symlink To `content://` URI 🫩.

**Comment by musty20580-tech** [comment]
Thanks for the update
